go/store/types/value_store.go

```diff
@@ -658,6 +670,48 @@ func (lvs *ValueStore) GC(ctx context.Context) error {
 	return nil
 }
 
+func (lvs *ValueStore) gcProcessRefs(ctx context.Context, visited hash.HashSet, toVisit []hash.HashSet, keepHashes func(hs []hash.Hash) error, walker *parallelRefWalker, oldGen, dest chunks.ChunkStore) error {
+	if len(toVisit) != 1 {
+		panic("Must be one initial hashset to visit")
+	}
+
+	toVisitCount := len(toVisit[0])
+	for toVisitCount > 0 {
+		batches := makeBatches(toVisit, toVisitCount)
+		toVisit = make([]hash.HashSet, len(batches))
+		toVisitCount = 0
+		for i, batch := range batches {
+			if err := keepHashes(batch); err != nil {
+				return err
+			}
+
+			vals, err := lvs.ReadManyValues(ctx, batch)
+			if err != nil {
+				return err
+			}
+			if len(vals) != len(batch) {
+				return errors.New("dangling reference found in chunk store")
+			}
+
+			hashes, err := walker.GetRefSet(visited, vals)
+			if err != nil {
+				return err
+			}
+
+			if oldGen != nil {
+				hashes, err = oldGen.HasMany(ctx, hashes)
```

Comment:
I would change the interface here and make it the responsibility of the caller to pass in something that could filter hashes as we're talking them to short circuit the walk. GenerationalChunkStore or NBSStore with a GenerationalChunkStore would end up passing an implementation of that thing that basically did `oldGen.HasMany`, but this layer wouldn't explicitly know about old gen.